### PR TITLE
Implementation for HRINT-1503

### PR DESCRIPTION
**Pull request by myarichuk**
*  Introduce single-threaded execution of tests with deterministic alphabetic ordering for DEBUG, VALIDATE or TEST_PERFORMANCE compilation variables
* wire test method beginning and ending via xUnit to add snapshots of memory/cpu via TestResourcesAnalyzer
* add perf test assembly runner to FastTests, SlowTests and RachisTests

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -122,3 +122,5 @@ bench/Micro.Benchmark/BenchmarkDotNet.Artifacts/
 /src/Raven.Pal/osxcross
 /src/Raven.Pal/*.nativecodeanalysis.xml
 /src/Raven.Pal/src/posix/*.o
+/src/Raven.Pal/Debug
+/src/Raven.Pal/build-all-posix.log
```
Comment (ppekrol) [inline .gitignore:126]: Is this needed?

File: test/SlowTests/SlowTests.csproj [hunk 2]
```diff
@@ -32,6 +32,7 @@
     <None Remove="Data\RavenDB_13940\storage-with-reused-journal-and-synced-data.zip" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="..\Tests.Infrastructure\CommonTestsAssemblyInfo.cs" Link="Properties\CommonTestsAssemblyInfo.cs" />
```
Comment (ppekrol) [inline test/SlowTests/SlowTests.csproj:35]: Don't forget about Embedded

File: test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using Xunit;
+
+#if DEBUG || VALIDATE || TEST_PERFORMANCE
```
Comment (ppekrol) [inline test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs]: only `TESTING`

File: test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+[assembly: TestCaseOrderer("Tests.Infrastructure.AlphabeticTestsOrderer", "Tests.Infrastructure")]
+[assembly: TestCollectionOrderer("Tests.Infrastructure.AlphabeticTestsOrderer", "Tests.Infrastructure")]
+[assembly: CollectionBehavior(DisableTestParallelization = true)]
```
Comment (ppekrol) [inline test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs]: Let's control that via xunit.setting.json

File: test/Tests.Infrastructure/AlphabeticTestsOrderer.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+using DiagnosticMessage = Xunit.Sdk.DiagnosticMessage;
+
+namespace Tests.Infrastructure
```
Comment (ppekrol) [inline test/Tests.Infrastructure/AlphabeticTestsOrderer.cs:8]: Move everything under one namespace

File: test/Tests.Infrastructure/TestResourcesAnalyzer.cs [hunk 6]
```diff
@@ -33,6 +35,21 @@ static TestResourcesAnalyzer()
...
+                _resources.Enqueue(TestResources.Create(GetDisplayName(testMethod), nameof(Start)));
+        }
+
+        internal static void End(ITestMethod testMethod)
+        {
+            if (_enabled)
+                _resources.Enqueue(TestResources.Create(GetDisplayName(testMethod), nameof(End)));
+        }
+
         internal static void Start(Context context)
```
Comment (ppekrol) [inline test/Tests.Infrastructure/TestResourcesAnalyzer.cs]: Need a cleanup here. We do not want to keep all the methods. Also we want to make it synchronous.

File: test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        {
+        }
+
+        protected override Task<RunSummary> RunTestMethodAsync(ITestMethod testMethod, IReflectionMethodInfo method, IEnumerable<IXunitTestCase> testCases, object[] constructorArguments)
+        {
+            TestResourcesAnalyzer.Start(testMethod);
+            return base.RunTestMethodAsync(testMethod, method, testCases, constructorArguments)
+                       .ContinueWith(t =>
+                       {
+                           TestResourcesAnalyzer.End(testMethod);
```
Comment (ppekrol) [inline test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs]: Remember that we want to gather 2 results here. Before and after GC.

File: .gitignore (hunk 1 above)
Comment (myarichuk) [reply .gitignore:126]: No. I will delete those, it should be in .gitignore

File: test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs (hunk 3 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs]: Ok. I will remove the DEBUG from there once everything else will be in place. No?

File: test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs (hunk 4 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs]: I have removed the DisableTestParallelization, as we discussed.

File: test/Tests.Infrastructure/TestResourcesAnalyzer.cs (hunk 6 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/TestResourcesAnalyzer.cs]: I will refactor this to be sync

File: test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs (hunk 7 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs]: Done (in the next commit)

File: test/Tests.Infrastructure/TestResourcesAnalyzer.cs (hunk 6 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/TestResourcesAnalyzer.cs]: Done (in the next commit)

File: test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            if(_resourceSnapshotEnabled)
+                _testResourceSnapshotWriter.WriteResourceSnapshot(TestStage.TestStarted, GetDisplayName(testMethod));
+
+            return base.RunTestMethodAsync(testMethod, method, testCases, constructorArguments)
+                       .ContinueWith(t =>
+                       {
+                           if(_resourceSnapshotEnabled)
+                               _testResourceSnapshotWriter.WriteResourceSnapshot(TestStage.TestFinishedBeforeGc, GetDisplayName(testMethod));
+
+                           GC.Collect(2, GCCollectionMode.Forced, true, true);
```
Comment (ppekrol) [inline test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs]: Do we want to compact LOH?
Also do we need to GC.WaitForPendingFinalizers?

**Review by ppekrol — COMMENTED** [review]
LGTM

File: test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs (hunk 8 above)
Comment (myarichuk) [reply test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs]: I think yes. Added that in the next commit.

**Comment by myarichuk** [comment]
Added an implementation of the service, it should do inserts to InfluxDB. 
* Note the **readme.txt** - for convenience, it contains docker commands to set up local environment (influx + influx UI + grafana)
* Note the *InfluxTag* attribute on the row object - it allows InfluxDB to group-by on that field

**Pull request closed without merging**